Feature/initial fat32

DrDeano opened the pull request:
This initial FAT32 filesystem will parse a stream, which could be a HDD or a in memory filesystem and produce a FAT32 filesystem. This will check if the raw bytes is a valid FAT32 filesystem as described in the spec.

This is only a skeleton with no open, read or write functionality.

Part of #216

p.s. have fun, this is a long one, lots more to come in later PR :P

SamTebbs33 commented:
Sorry, I had a quick look a couple of days ago but got distracted. I'll have a proper look tomorrow :)

SamTebbs33 commented on src/kernel/code_page.zig:
```diff
@@ -0,0 +1,88 @@
+    ///     error.InvalidChar - The character to be converted is not in the code page table.
+    ///
+    pub fn toCodePage(code_page: CodePages, char: u16) Error!u8 {
+        const table = switch (code_page) {
+            .CP437 => cp437.table,
+        };
```
This could be made nice and abstract by adding a function to `CodePages` that returns the table. This could then be used in `toWideChar` as well.

DrDeano replied:
Not really sure what you mean. is it like:
```Zig
pub fn getTable(code_page: CodePage) []u8 {
    return switch (code_page) {
            .CP437 => cp437.table,
        };
}
```

SamTebbs33 replied:
Exactly. That could go in the struct.

DrDeano replied:
what would the use be of it though? but can add it

SamTebbs33 replied:
Well you have this logic in a couple of places so it would be good to extract it into the enum itself. Plus, encapsulation is nice :)

DrDeano replied:
I see now

SamTebbs33 commented on src/kernel/code_page.zig:
```diff
@@ -0,0 +1,88 @@
+const std = @import("std");
```
Since this is more of a supporting file, it could be moved to src/kernel/code_page

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+/// The boot record for FAT32. This is use for parsing the initial boot sector to extract the
+/// relevant information for todays FAT32.
+const BootRecord = struct {
+    /// The jump bytes that begin the boot code. This will jump over the FAT32 header.
+    jmp: [3]u8,
```
Do we need to keep this around if the disk is only used for storage and not booting?

DrDeano replied:
This is only here for parsing the boot record. This struct isn't used in the main FAT32FS.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+
+    /// The size in bytes of the root directory. This is only used by FAT12 and FAT16, so is
+    /// always 0 for FAT32.
+    zero_root_directory_size: u16,
```
The `zero` part of the name is a bit confusing. I think that documentation saying that it's 0 for fat32 is enough. Same for the other fields with `zero` in their name.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+    /// timestamp of creation.
+    serial_number: u32,
+
+    /// The partitioned volume label.
+    volume_label: [11]u8,
+
+    /// The file system type.
```
I think it would be nice to have some documentation as to why these are arrays.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+    struct_signature: u32,
+
+    /// The number of free clusters in the image
+    number_free_cluster: u32,
```
cluster**s**?

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+    ///     Whether the file is a directory or not.
+    ///
+    pub fn isDir(self: *const ShortName) bool {
+        return self.attributes & @enumToInt(ShortName.Attributes.Directory) == @enumToInt(ShortName.Attributes.Directory);
```
`@enumToInt(ShortName.Attributes.Directory) == @enumToInt(ShortName.Attributes.Directory);` will never be false.

DrDeano replied:
`self.attributes & @enumToInt(ShortName.Attributes.Directory)` will evaluate first then the equality. I have a test for this.

SamTebbs33 replied:
Oh of course, I was reading this after writing some C++ at work and read the `&` as `&&`! Perhaps some parentheses would be nice anyway.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+
+///
+/// Initialise a struct from bytes.
+/// TODO: Once packed structs are good to go, then use std.mem.bytesAsValue.
```
I've had success with packed structs in my ELF work, have you tried using them recently?

DrDeano replied:
I have did an update and when I try using packed structs, it doesn't compile and not error returned, just exit code 1. Plus the sizes are not expected. So will have to wait a bit longer.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+            self.opened_files.deinit();
+            self.allocator.destroy(self.root_node.node);
+            self.allocator.destroy(self.fs);
+            self.allocator.destroy(self);
```
In general I don't think deinit functions should destroy the struct itself since that would mean you couldn't deinit stack-allocated structs.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+        ///                       filesystem. See Error for the list of possible errors.
+        ///
+        pub fn init(allocator: *Allocator, stream: StreamType) (Allocator.Error || ReadError || SeekError || Error)!*Fat32Self {
+            log.info("Init\n", .{});
+            defer log.info("Done\n", .{});
```
Fat32 initialisation will be happening after boot initialisation too, so I don't think these log messages are appropriate. We don't want them appearing when the user plugs in a memory stick, for example (we'd want device related log messages instead).

DrDeano replied:
Shall i change it to debug logs

SamTebbs33 replied:
Yeah that's a good idea

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+/// Error: Allocator.Error || Fat32FS(@TypeOf(stream)).Error
+///     Allocator.Error - If there isn't enough memory to create the filesystem.
+///
+pub fn initialiseFAT32(allocator: *Allocator, stream: anytype) (Allocator.Error || ErrorSet(@TypeOf(stream)) || Fat32FS(@TypeOf(stream)).Error)!*Fat32FS(@TypeOf(stream)) {
```
Do we need this when we have Fat32FS.init?

DrDeano replied:
It just makes the API easier to use. Plus this is how I have seen it in the std lib.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+    return Fat32FS(@TypeOf(stream)).init(allocator, stream);
+}
+
+/// The number of allocations that the init function make.
+const init_allocations: usize = 5;
+
+/// The test buffer for the test filesystem stream.
+var stream_buff: [if (builtin.is_test) 34090496 else 0]u8 = undefined;
```
It would be good to prefix these variables used for testing with `test`.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -0,0 +1,1756 @@
+        "/dot.in.file.txt",
+        "/file.long_ext",
+        "/file.t x t",
+        "/insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long_insanely_long.txt",
```
nice xD

DrDeano replied:
If you can think of other test files, let me know

SamTebbs33 reviewed (changes requested):
Looking good! Nice job on getting so much done.

SamTebbs33 reviewed (approved):
:rocket:

DrDeano committed:
Initial FAT32 interface

Include code page 437

This is now moved to a better location

Fixed initrd doc for allocator


Added test files

 Plus minor extras

Created initial FAT32 filesystem

This initial FAT32 filesystem will parse a stream, which could be a HDD or a in memory filesystem and produce a FAT32 filesystem. This will check if the raw bytes is a valid FAT32 filesystem as descibed in the spec.

This is only a skeleton with no open, read or write functionality.

Part of #216

logger -> log


Fmt


Removed 'zero_' from struct fields


number_free_cluster -> number_free_clusters


Return non allocated FAT32FS


prefix test_


Added doc comment for boot sector arrays


Moved code_page.zig


info -> debug for init FAT32


Abstracted getting code page table
